src/pudl/etl/__init__.py

```diff
@@ -209,13 +209,18 @@ def _get_keys_from_assets(
     "epacems_io_manager": epacems_io_manager,
 }
 
-# By default, limit CEMS year processing concurrency to prevent memory overload.
+# By default, limit CEMS and FERC 714 processing concurrency to prevent memory overload.
 default_tag_concurrency_limits = [
     {
         "key": "datasource",
         "value": "epacems",
         "limit": 2,
-    }
+    },
+    {
+        "key": "datasource",
+        "value": "ferc714",
+        "limit": 1,
+    },
```

Comment:
I don't *think* this will have any effect. The FERC-714 job is parallelized internally, not by virtue of the Dagster infrastructure. Right now it will use all available CPUs no matter what. I think it's the function deep down inside that needs to be told to do something different. By contrast the CEMS work units *are* being managed by Dagster.

Reply:
* If I run `_out_ferc714__hourly_demand_matrix` by itself, it uses 1 CPU and ~4GB of memory, and takes ~3 minutes to complete.
* And then `_out_ferc714__hourly_imputed_demand` is the asset that tries to use lots of cores, but all within a single asset (the parallelism has nothing to do with Dagster). It's often running at 4-600% (on my laptop) with one process that seems to need ~1-2 GB of memory.

Reply:
Ah okay, it's the `numpy.linalg.svd` and `numpy.einsum` functions which rely on multithreaded BLAS and LAPACK libraries, whose behavior is controlled by environment variables or other libraries...

https://numpy.org/doc/stable/reference/routines.linalg.html

Reply:
But again the uses-lots-of-cores asset doesn't seem to be a uses-lots-of-memory asset so I'm confused.

Reply:
`_out_ferc714__hourly_imputed_demand` wasn't running when we got the OOM error though. Assuming `_out_ferc714__hourly_demand_matrix` pushed us over the memory edge I thought we could limit the number of processes running when the asset is run. Maybe this is just playing wack a mole though.

I should probably do a full memory profile locally to understand what the culprit is.

Reply:
Oh is this a *global* concurrency limit? Like if this process is running, then no other Dagster ops can be running?

Reply:
Or are we also applying a global concurrency limit on the CEMS assets? It always seemed like other stuff was running, but only 2 CEMS assets would ever be materializing at a time.  If we have a global concurrency limit on CEMS we could probably run more than 2 at a time. I think each of them takes ~2GB of memory peak.

Reply:
I also think that these are *tag* level concurrency and that @e-belfer is right. I think we have a couple options:

1. make a `high-mem` op_tag, set its tag concurrency to $LOW_NUM
2. set `max_concurrent` (which is the *global* concurrency limit) to $MEDIUM_NUM

We could combine these, also - say our highest memory usage asset uses 30GB memory, and our `high-mem` threshold is ~8GB. Then, if the `high-mem` concurrency limit is 1, we could set the *global* concurrency limit to 5 and still expect our peak memory usage to be less than (30 * 1 + 8 * 4) = 62 GB. 

We don't have to have an exhaustively-correct `high-mem` tag either - I think just picking off a few culprits will improve our reliability here.

Finally, it's maybe worth looking at some of [these recommendations](https://github.com/dagster-io/dagster/discussions/17616) for reducing memory usage - but that seems like more effort than the `high-mem` bandaid.

Reply:
I like the generic nature of the `high-mem` tag, at least as a short-term fix.

Probably a lot of our timeseries processing could be parallelized thoughtfully with Dask, and it seems like there could be a lot more timeseries in our future. But it'll take much more specific effort.

If/When we switch to using a Dask cluster as the executor (or anything other than One Big Machine) is there a way to use something like the `high-mem` tag to make sure that the node that gets that job has enough memory, without requiring *every* node to have a lot of memory?

Reply:
I also like the `high-mem` tag idea. Would it protect us against a scenario where we have one `high-mem` asset running and a bunch of other medium - low memory assets running?

We could try just using the `high-mem` tag and if it doesn't work then we can limit global concurrency.

Reply:
> If/When we switch to using a Dask cluster as the executor (or anything other than One Big Machine) is there a way to use something like the `high-mem` tag to make sure that the node that gets that job has enough memory, without requiring _every_ node to have a lot of memory?

Yeah, it looks like Dask supports differently-resourced worker types/pools: https://blog.dask.org/2022/02/17/helm-multiple-worker-groups

> I also like the high-mem tag idea. Would it protect us against a scenario where we have one high-mem asset running and a bunch of other medium - low memory assets running?
 
Not really, if we have 64 GB and our 30GB task is running, there's nothing to stop dagster from scheduling 1000 1GB tasks also. So the global limit is necessary to fully protect us.